server/service/datadeal/service_data_validation_test.go

```diff
@@ -0,0 +1,337 @@
+	"github.com/cosmos/cosmos-sdk/crypto/keys/secp256k1"
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	datadealtypes "github.com/medibloc/panacea-core/v2/x/datadeal/types"
+	"github.com/medibloc/panacea-oracle/crypto"
+	"github.com/medibloc/panacea-oracle/mocks"
+	"github.com/medibloc/panacea-oracle/panacea"
+	datadeal "github.com/medibloc/panacea-oracle/pb/datadeal/v0"
+	"github.com/medibloc/panacea-oracle/server/rpc/interceptor/auth"
+	"github.com/medibloc/panacea-oracle/server/service/key"
+	"github.com/medibloc/panacea-oracle/sgx"
+	"github.com/stretchr/testify/suite"
+)
+
+type dataDealServiceServerTestSuite struct {
+	suite.Suite
+
+	svc *mocks.MockService
+
+	productID []byte
+	uniqueID  []byte
+
+	oracleAcc     *panacea.OracleAccount
+	oraclePrivKey *btcec.PrivateKey
+
+	deal *datadealtypes.Deal
+
+	providerAccPrivKey secp256k1.PrivKey
+	providerAccPubKey  cryptotypes.PubKey
+	providerAcc        authtypes.AccountI
+}
+
+func TestDataDealServiceServer(t *testing.T) {
+	suite.Run(t, &dataDealServiceServerTestSuite{})
+}
+
+func (suite *dataDealServiceServerTestSuite) BeforeTest(_, _ string) {
+	suite.productID = []byte("productID")
+	suite.uniqueID = []byte("uniqueID")
+
+	mnemonic, _ := crypto.NewMnemonic()
+	suite.oracleAcc, _ = panacea.NewOracleAccount(mnemonic, 0, 0)
+	suite.oraclePrivKey, _ = btcec.NewPrivateKey(btcec.S256())
+
+	suite.deal = &datadealtypes.Deal{
+		Id:         1,
+		DataSchema: []string{"https://json.schemastore.org/github-issue-forms.json"},
+		Status:     datadealtypes.DEAL_STATUS_ACTIVE,
+	}
+	suite.providerAccPrivKey = *secp256k1.GenPrivKey()
+	suite.providerAccPubKey = suite.providerAccPrivKey.PubKey()
+	suite.providerAcc = mocks.NewMockAccount(suite.providerAccPubKey)
+	suite.svc = &mocks.MockService{
+		QueryClient: &mocks.MockQueryClient{
+			Account: suite.providerAcc,
+			Deal:    suite.deal,
+		},
+		Ipfs:          &mocks.MockIPFS{},
+		EnclaveInfo:   sgx.NewEnclaveInfo(suite.productID, suite.uniqueID),
+		OracleAccount: suite.oracleAcc,
+		OraclePrivKey: suite.oraclePrivKey,
+	}
+}
+
+func (suite *dataDealServiceServerTestSuite) AfterTest(_, _ string) {
+	mocks.RemoveMockIPFSData()
+}
+
+func (suite *dataDealServiceServerTestSuite) TestValidateDataSuccess() {
+	// provide data
+	jsonDataBz := []byte(
+		`
+		{
+			"name": "name",
+			"description": "description",
+			"body": [{ "type": "markdown", "attributes": { "value": "val1" } }]
+		}
+		`)
+
+	// encrypted provider data with provider private key and oracle public key
+	providerPrivKey, _ := btcec.PrivKeyFromBytes(btcec.S256(), suite.providerAccPrivKey.Bytes())
+
+	sharedKey := crypto.DeriveSharedKey(
+		providerPrivKey,
+		suite.oraclePrivKey.PubKey(),
+		crypto.KDFSHA256,
+	)
+
+	encryptedData, err := crypto.Encrypt(sharedKey, nil, jsonDataBz)
+	suite.Require().NoError(err)
+
+	dataHash := sha256.Sum256(jsonDataBz)
```

Comment:
Although, this isn't the part of your PR, let's put a TODO comment somewhere, which says that we have to serialize JSON data canonically, as Tendermint does for serialize tx body before signing on it. 

The canonical serialization, such as sorting JSON fields in alphabetical order, is very crucial to produce the same hash values from the semantically same JSON data.

Reply:
As I remember, VC/VP also does the canonical serialization to produce its proof.

Reply:
Yeah, that's a very important point.
I've been thinking about that as well.